test/model/internal_link_test.dart

```diff
@@ -0,0 +1,486 @@
+      // (true, 'same domain, punycoded host',
+      //   Uri.parse('https://example.xn--h2brj9c/#narrow/stream/check'),
+      //   Uri.parse('https://example.भारत/')), // FAILS
+
+      (true, 'punycodable host',
+        Uri.parse('https://example.भारत/#narrow/stream/check'),
+        Uri.parse('https://example.भारत/')),
+
+      // (true, 'same domain, IDNA-mappable',
+      //   Uri.parse('https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/check'),
+      //   Uri.parse('https://example.com')), // FAILS
+
+      (true, 'ipv4 address',
+        Uri.parse('http://192.168.0.1/#narrow/stream/check'),
+        Uri.parse('http://192.168.0.1/')),
+
+      // (true, 'same IPv4 address, IDNA-mappable',
+      //   Uri.parse('http://１𝟗𝟚。①⁶🯸．₀｡𝟭/#narrow/stream/check'),
+      //   Uri.parse('http://192.168.0.1/')), // FAILS
+
+      // TODO: Add tests for IPv6.
+
+      // These examples may seem weird, but a previous version of
+      // the zulip-mobile code accepted most of them.
+
+      // This one, except possibly the fragment, is a 100% realistic link
+      // for innocent normal use.  The buggy old version narrowly avoided
+      // accepting it... but would accept all the variations below.
+      (false, 'wrong domain, realm-like path, narrow-like fragment',
+        Uri.parse('https://web.archive.org/web/*/${realmUrl.resolve('#narrow/stream/check').toString()}'),
+        realmUrl),
+      (false, 'odd scheme, wrong domain, realm-like path, narrow-like fragment',
+        Uri.parse('ftp://web.archive.org/web/*/${realmUrl.resolve('#narrow/stream/check').toString()}'),
+        realmUrl),
+      (false, 'same domain, realm-like path, narrow-like fragment',
+        realmUrl.resolve('web/*/${realmUrl.resolve('#narrow/stream/check').toString()}'),
+        realmUrl),
+    ];
+    for (final testCase in testCases) {
+      final bool expected = testCase.$1;
+      final String description = testCase.$2;
+      final Uri url = testCase.$3;
+      final Uri realmUrl = testCase.$4;
+      final String testName = '${expected ? 'accept': 'reject'} $description: ${url.toString()}';
+      test(testName, () {
+        final store = setupStore(realmUrl: realmUrl, streams: streams);
+        final result = parseInternalLink(url, store);
+        check(result != null).equals(expected);
+      });
+    }
+  });
+
+  group('General link parsing', () {
+    final streams = [
+      eg.stream(streamId: 1, name: 'check'),
+      eg.stream(streamId: 3, name: 'mobile'),
+      eg.stream(streamId: 5, name: 'stream'),
+      eg.stream(streamId: 123, name: 'topic'),
+    ];
+
+    test('/#narrow/stream/... returns expected StreamNarrow', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('/#narrow/stream/check', const StreamNarrow(1)),
+        ('/#narrow/stream/stream/', const StreamNarrow(5)),
+        ('/#narrow/stream/topic/', const StreamNarrow(123)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
```

Comment:
I've found out that it's okay to have multiple layers of groups, so I've refactored the tests that each test case becomes its own `test()` call.

Reply:
Great, thanks.

And yeah, in general, if there's something that was asked for (in a review, or in the issue description, etc.) that doesn't seem to have been possible to do, please be sure to flag it explicitly. That way we can discuss alternatives and tradeoffs — or perhaps someone will know a way it can be done after all.